### DISCUSSION TITLE
Add lifted operator support for nullable decimals

### DISCUSSION THREAD

**Discussion by evolvedlight** [dotnet/csharplang#3973] [Language Ideas] [answered]
Hi,

I have a half-baked proposal, but maybe it'd be too much of a breaking change.

Given the following code:
```
decimal? a = 2;
decimal? b = null;
var c = (a + b);
```
I think that many people might expect the answer to be 2, with type nullable decimal. Instead as there's no lifted operator it returns null.

Could this behavior be considered, or is this too much of a change to the language at this point? 

Thanks,
Steve

**Comment by HaloFour** [comment, marked as answer]
This is the case with all of the nullable types and it is on purpose.  `null` is not the same as `0`, it's treated more like "unknown", like NULL in SQL.  Changing that would be a breaking change.

**Reply by evolvedlight** [reply]
Thanks. Seems however that the same decision wasn't made with Linq - if `a` and `b` are in a list and you call `.Sum()` then the opposite of this rule is used:
```
(new List<decimal?> { a, b }).Sum()
2
```
So there's one rule for addition, and the opposite for Sum in Linq, and that seems like a weird design.

**Reply by HaloFour** [reply]
I don't think I noticed that before but it doesn't surprise me.  When nullable value types were added in .NET 2.0 there wasn't really a prescribed way for handling binary operators between them.  C# and VB.NET differed in their implementations from the outset.  As LINQ was introduced later in .NET 3.5 as an API it had to choose an implementation that would apply across all languages.

**Reply by jnm2** [reply]
`new decimal?[] { null, 2 }.Sum()` is also consistent with `select sum(N) from (values (null), (2)) as SomeValues (N)`. Both evaluate to 2.

**Comment by huoyaoyuan** [comment]
> I think that many people might expect the answer to be 2

```csharp
> int? a = 1;
> int? b = null;
> a + b
null
```

The lifting rule is consistent among types. Arithmetic operation for any `null` operand produces `null`.

**Comment by spydacarnage** [comment]
Isn't that still the same in SQL?

1 + null = null, but SUM(column) where column = nullable int ignores nulls ?
